### PR TITLE
Eliminate silent failures in CI.

### LINKED ISSUE #241: CI fails silently
I am sorry to report this:

CI is currently failing silently.  Turns out that papering over the memory issue (can be undone with #216) was a bad idea, or at least needs tweaking.  Looking down through the CI capture for `pytest` sections, I saw a failing test for CTI file reads, but a green checkmark on the CI dashboard.

The CTI file read issue appears to be fixed (by #239), so it is no longer a good example, unless we can dig up CI history on it.

The fix for CI may be as simple as #240, or similar. Either way, paper can only go so far and it is tearing.

### PR DESCRIPTION
**Pull request by MTCam**
Trying to get CI to speak up when individual `pytest` invocation has failing tests.  Individual `pytest` running was introduced to paper over a memory issue (#212).  The papering can be undone with #216.

The current example where CI fails silently is in [this CI run](https://github.com/illinois-ceesd/mirgecom/runs/1860379003?check_suite_focus=true) of #239. Unfortunately, this is already happening in master as a test in #232 causes a failure, but CI dashboard has a green checkmark.

* The failing test from #232 is now fixed by #239.
* _This_ PR (#240) successfully gets CI to fail when individual files are tested with `pytest`. It should be merged to keep the paper-over working properly. 
* The CI for #240 should only pass when all the tests in master are passing

Fixes #241

### CODE REVIEW

**Comment by matthiasdiener** [comment]
Why not just `python ... || exit 1`?

**Comment by majosm** [comment]
Silent CI failures?

**Comment by MTCam** [comment]
Kind of strange... If it *fails* that means it's working, and is ready for merge.
:thinking:

**Comment by majosm** [comment]
May want to do `exit $?` so we get the right error code.

This still seems weird to me... I don't get why the explicit `exit` should be necessary.

**Comment by matthiasdiener** [comment]
> May want to do `exit $?` so we get the right error code.

👍 

> This still seems weird to me... I don't get why the explicit `exit` should be necessary.

I think because the CI checks the return code of the whole script. Bash usually just continues executing even when an error occurs midway through the script. (this is why options such as `set -o errexit` exist)

**Comment by majosm** [comment]
Ohhh. That's nasty. Is it possible/advisable to do `set -e`?

**Comment by inducer** [comment]
> Is it possible/advisable to do `set -e`?

- IMO: That should always be the first line of any shell script.
- AFAIK, Github CI runs its scripts with `set -e` by default.

**Comment by matthiasdiener** [comment]
> Ohhh. That's nasty. Is it possible/advisable to do `set -e`?

I think that's risky. Sometimes you want it to continue on failure. E.g., the following does not work:

```bash
set -o errexit

ret=$(false)

# Not executed:
echo $ret
```

**Comment by matthiasdiener** [comment]
> * AFAIK, Github CI runs its scripts with `set -e` by default.

[Yes](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#using-a-specific-shell), but we override it.

**Comment by majosm** [comment]
> I think that's risky. Sometimes you want it to continue on failure. E.g., the following does not work:

I think when you want to continue on failure you can do:

```bash
<command> || result=$?

echo $result
```

Exit on error seems like a safer default to me.

**Comment by matthiasdiener** [comment]
> > I think that's risky. Sometimes you want it to continue on failure. E.g., the following does not work:
> 
> I think when you want to continue on failure you can do:
> 
> ```shell
> <command> || result=$?
> 
> echo $result
> ```
> 
> Exit on error seems like a safer default to me.

Sure, but you might want to run external scripts that rely on the default error behavior.

**Comment by matthiasdiener** [comment]
> > Exit on error seems like a safer default to me.
> 
> Sure, but you might want to run external scripts that rely on the default error behavior.

Just to be clear, I am not against setting -e on these CI scripts, just that the rules for errexit are really arcane (see https://stackoverflow.com/questions/44819831/bash-errexit-with-arithmetic-expansion for a fun one) and not very reliable.

**Review by matthiasdiener — APPROVED** [review]
(cc #239)

**Comment by MTCam** [comment]
Which should be merged first?  My current plan would be to do #239 firs, then #240 should start passing, and I'll merge it. Do we need further discussion or final review @inducer?

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -91,7 +91,7 @@ jobs:
             # One test at a time to address ballooning memory.
             # https://github.com/illinois-ceesd/mirgecom/issues/212
             for filename in test_*.py ; do
-              python -m pytest --durations=10 --tb=native --junitxml=pytest.xml --doctest-modules -rxsw ./"$filename"
+              python -m pytest --durations=10 --tb=native --junitxml=pytest.xml --doctest-modules -rxsw ./"$filename" || exit $?
```
Comment (inducer) [inline .github/workflows/ci.yaml:94]: > > * AFAIK, Github CI runs its scripts with `set -e` by default.
> 
> [Yes](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#using-a-specific-shell), but we override it.
> 
> Edit: here is where we override that:
> 
> https://github.com/illinois-ceesd/mirgecom/blob/d5c6b0ddf8da1778268e13df169174216a5aef3c/.github/workflows/ci.yaml#L82-L83

My strong preference would be to undo the shell override instead and leave the default of `set -e` in place.

**Review by inducer — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (MTCam) [reply .github/workflows/ci.yaml:94]: Wanna take a crack at this @matthiasdiener?

Comment (MTCam) [reply .github/workflows/ci.yaml]: Successfully failed. :tada:

Comment (matthiasdiener) [reply .github/workflows/ci.yaml]: its rare to be so happy about a failing CI run :-D

**Comment by MTCam** [comment]
I guess our changes to the CTI file reader produced another non-working version:
```
_______________________________ test_cti_reader ________________________________
Traceback (most recent call last):
  File "/home/runner/work/mirgecom/mirgecom/test/test_cti.py", line 34, in test_cti_reader
    first_line, _, _ = test_cti.partition("\n")
TypeError: a bytes-like object is required, not 'str'
--- generated xml file: /home/runner/work/mirgecom/mirgecom/test/pytest.xml ----
============================= slowest 10 durations =============================
```

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -36,7 +36,7 @@ jobs:
           with:
             python-version: '3.x'
         - name: Install
-          shell: bash -l {0}
+          shell: bash -e -l {0}
```
Comment (inducer) [inline .github/workflows/ci.yaml:39]:
```suggestion
```
Or, put, differently, what's the advantage to having this? (Sorry, we may have had this discussion. I just don't remember.)

Comment (matthiasdiener) [reply .github/workflows/ci.yaml:39]: I dont remember either. For some reason, we must have thought of using a login shell here.

Comment (inducer) [reply .github/workflows/ci.yaml:39]: Kill it?

**Review by inducer — APPROVED** [review]

**Pull request merged**